htdocs/compta/facture/class/facture.class.php

```diff
@@ -1574,11 +1574,19 @@ public function fetch($rowid, $ref = '', $ref_ext = '', $notused = '', $fetch_si
 				$this->close_code			= $obj->close_code;
 				$this->close_note			= $obj->close_note;
 
-				$this->socid = $obj->fk_soc;
-				$this->thirdparty = null; // Clear if another value was already set by fetch_thirdparty
+				if ($this->socid != $obj->fk_soc) {
```

Comment:
For consistency, we try to have this philosophy for fetch:
It loads only data on the "record" level (so properties and extrafields because it is a 1-1 relation) and the lines on object. All other subobject (thirdparty, project, ...) that are object must be loaded after the fetch when needed.

So can you try to fix by adding instead a fetch_thirdparty after the fetch where it misses.

Reply:
For your question, what will be goal of CommonTransaction

A good blanced choice would be:

CommonObject 
    CommonInvoice
      Invoice
      SupplierInvoice
    CommonProposal
      Proposal
      SupplierProposal
    CommonOrder
      Order
      SupplierOrder

Reply:
> For consistency, we try to have this philosophy for fetch:
> It loads only data on the "record" level (so properties and extrafields because it is a 1-1 relation) and the lines on object. All other subobjects (thirdparty, project, user of creation, user of modification, etc...) that are object must be loaded after the fetch (so outside the fetch), and only when needed.
> 
> So can you try to fix by adding instead a fetch_thirdparty after the fetch where it misses. We have more and more objects, and loading them always any time is an important overload we should avoid.

I see what you mean and actually I expected an answer like this. However, what I was doing here is to check if the "subobject" was loaded before and only if it was and if any IDs have changed then load it again.
If I'd do it like you suggested then all subobject values would be fetched again no matter if they were changed or not. So the way I suggested should be more efficient than fetching "subobjects" like thirdparty or project again and again even if these subobjects haven't been changed.
Looking into the code I found so many locations where `someobject->fetch()` is being called and it is expected that - if subojbects like thirdparty or project were fetched before - these objects won't suddenly be cleared and can be used as before. But since 12.0.0 subobjects like third_party or project are being set to null, which results in a lot of problems.
Additionally I don't think it is good coding to **copy** a pair of lines (`someobject->fetch(); someobject->fetch_thirdparty();`) again and again all over the code. At the end there will always be programmers who just forget to call some "subojbect fetching" methods. Especially if...

...if there will become more subobjects: you would have to copy code lines like `someobject->fetch(); someobject->fetch_thirdparty(); someobject->fetch_project(); someobject->fetch_othersubobject();` ... all over the code. 

For this reason I suggested to create another class "CommonTransaction" to benefit from OO advantages like inheritance or virtual methods. Please read my answer regarding your other question about a "CommonTransaction" class suggestion (which will follow in a few minutes).

However, your recommendation might work for the moment. I would really check the code and add another line `someobject->fetch_thirdparty();` everywhere where `someobject->fetch();` appears and where it _seems_ that `someobject->fetch_thirdparty();` is needed because it was called before _(though I really don't like this way and think this will make the code less maintainable)_.

Reply:
> For your question, what will be goal of CommonTransaction (if we need common method, it is goal of method CommonObject).
> 
> A good balanced choice would be:
> 
> ```
> CommonObject 
>     CommonInvoice
>       Invoice
>       SupplierInvoice
>     CommonProposal
>       Proposal
>       SupplierProposal
>     CommonOrder
>       Order
>       SupplierOrder
> ```

The reason I thought about an additional class "CommonTransaction" is that there is some common objectoriented philosophy with "CommonObject", but in my opinion, it is _too_ common for _all_ Dolibarr objects.

There are objects like companies, contacts, products, holidays, invoices, proposals, orders and so on. They all use CommonObject, which makes sense regarding their common properties like "how to read any data from database" or other stuff. However, objects like invoices don't have much in common with products or holidays. Products have properties like size, weight, price where invoices have properties like lines, sum, third party etc.
But orders are very similar to proposals, invoices or - I call them - other (business-)transactions (IIRC the [UBL-standard](https://en.wikipedia.org/wiki/Universal_Business_Language) uses the same terminology.)

So what I thought is, why not keeping all _common_ stuff (properties and methods) of these "transaction" classes in one common class and inherit them. This would make it unnecessary to copy common procedures (methods) to all transaction objects again and again. I.e. it makes it much more efficient to make changes or even fixes to the code.

Having methods like `CommonTransaction->fetch_thirdparty()` could then have a special logic to decide there is really a fetch necessary from the database again or not. Just for example you could decide if the object id (rowid) is still the same and only if there is a "need thirdparty update" flag set in the transaction object then really (and automatically) do a fetch of thirdparty from the database. However, logic like this should be implemented at a common location and not copied again and again over all the transaction classes.

Other methods like or `Proposal->add_product()` and `Order->add_product()` could then be at a single location in a class `CommonTransaction`.

Do you see, what I mean?

```
CommonObject (Common methods and properties for _all_ Dolibarr objects)
  CommonTransaction (Common methods and properties for transaction objects)
    CommonInvoice (Common methods and properties for incoices only)
      Invoice
      SupplierInvoice
    CommonProposal
      Proposal
...
```

Reply:
Just one additional information: in compta/facture/card.php, a call to `object->fetch();` is at list at 70 locations. Replacing them all with `object->fetch(); object->fetch_thirdparty();` wouldn't be a good idea, I think. IMHO having some central logic to decide if a fetch is necessary would be more efficient.

Reply:
There is not only fetch_thirdparty, there is also project, user that create, user that modify, etc...
If do it for one type of child, we should do it for others two or developer will have to enter core code each time a fetch is done to remember if child are loaded of not. Consistency is also important for maintainability.
And it hurting me to add code that ill solve problem when coder has forgot seems and and code that load data that we don't want when coder has done its work correctly. The good coder has the wrong effect and the bad coder ahs the good one. We expect to have the opposite, don't you think so ?

Reply:
No, I don't think this way.

The problem is: I don't know how to check and correct this issue in millions of locations in the code in a reasonable time. A central solution as propposed would be fine. As I mentioned before: With the solution that I suggested the object won't be loaded if it wasn't before.

IMHO, object oriented programming is not just giving simple structs names like CommonObject, but also make use of methods, inheritance, polymorphism and so on.

However, I really don't know how to fix this issue in so many locations. Maybe someone else should do this then.

Reply:
What is you rpoint of view about the magic method __get on property ->thirdparty ?
This may be a good solution to solve this trouble with best of 2 words (a centralized solution to load data and a loading only when we need it):  https://www.zentut.com/php-tutorial/php-magic-methods/

Reply:
Wow! I didn't expect this answer 😀

Actually I know these "magic methods", I love them and it would be the way **I would realize this**. The Dolibarr code, however, is not using technologies like this at all. Very rare inheritance, if at all, and technologies like polymorphism not at all, IIRC. Even access methods are very rare. Probably this is all by historical reasons. However, when trying to manage all this copy-paste code it is getting more and more difficult to keep everything manageable and running.

Actually, magic methods like __get and __set are being used in other OOP-languanges since decades. Delphi for example has access scopes to methods like private, protected, public and _published_. The latter is similar to these magic methods. PHP, however, wasn't a real object oriented language from the beginning.

So: Yes! I'd really love to consolidate properties like _thirdparty_ (but others, too) into parent classes and not repeat copying code again and again.

(And: I would recommend to create additional parent classes for objects which are similar one to each other like Proposal, Order or Invoice. I would place such classes between CommonObject and Invoice, Order, Proposal... Please see my comment above.)

Reply:
Yes, some things are not done for historical reason but also for compatibility reason. Are you ok to try a first PR to experiment to solve this issue using a __get property (as it is an experiment, we should try with this only). I have pushed a fixed for this issue in code for 13.0, but we can try to integrate a PR that will introduce use of magic method as soon as develop branch will be open again for new architectural change (soon i hope). It seems you know well the topic, so a first PR coming as an experiment before opening the door more widely would be appreciated. I suggest to open another PR however (mention to this ticket may be enough).

Reply:
Yes, I would really like to try a first PR to solve this (and maybe similar) issues.

Maybe we should continue further communication on this topic at a different location.